assignments/array-methods.js

```diff
@@ -2,82 +2,538 @@
-let largeShirts = [];
+// The large shirts won't be available for the event due to an ordering issue.
+// Get a list of runners with large sized shirts so they can choose a
+// different size. Return an array named largeShirts that contains information
+// about the runners that have a shirt size of L and log the result
+let largeShirts = runners.filter(item => item.shirt_size === 'L');
 console.log(largeShirts);
 
 // ==== Challenge 4: Use .reduce() ====
-// The donations need to be tallied up and reported for tax purposes. Add up all the donations into a ticketPriceTotal array and log the result
-let ticketPriceTotal = [];
+// The donations need to be tallied up and reported for tax purposes. Add up
+// all the donations into a ticketPriceTotal array and log the result
+let ticketPriceTotal = runners.reduce(
+    (total, item) => total + item.donation,
+    0
+);
 console.log(ticketPriceTotal);
 
 // ==== Challenge 5: Be Creative ====
-// Now that you have used .forEach(), .map(), .filter(), and .reduce().  I want you to think of potential problems you could solve given the data set and the 5k fun run theme.  Try to create and then solve 3 unique problems using one or many of the array methods listed above.
+// Now that you have used .forEach(), .map(), .filter(), and .reduce().  I
+// want you to think of potential problems you could solve given the data set
+// and the 5k fun run theme.  Try to create and then solve 3 unique problems
+// using one or many of the array methods listed above.
+
+// seperate runners into bronze, silver, and gold categories based on how many donations they got and then list out their names.
+let bronze = runners
+    .filter(item => item.donation >= 50 && item.donation < 100)
+    .map(item => `${item.first_name} ${item.last_name}`);
+// console.log(bronze);
+let silver = runners
+    .filter(item => item.donation >= 100 && item.donation < 200)
+    .map(item => `${item.first_name} ${item.last_name}`);
+// console.log(silver);
+let gold = runners
+    .filter(item => item.donation >= 200)
+    .map(item => `${item.first_name} ${item.last_name}`);
+// console.log(gold);
+
+// Sort students by shirt size
+
+let shirtSizes = {
+    L: [],
+    M: [],
+    S: []
+};
 
-// Problem 1
+shirtSizes['L'] = runners.filter(item => item.shirt_size === 'L');
+shirtSizes['M'] = runners.filter(item => item.shirt_size === 'M');
+shirtSizes['S'] = runners.filter(item => item.shirt_size === 'S');
+console.log(shirtSizes);
 
-// Problem 2
+// Problem 3
+// associate donation total to each company name.
 
-// Problem 3
\ No newline at end of file
+let donors = runners.map(item => item.company_name);
+donors.forEach(
+    company =>
+        (donors[company] = runners
+            .filter(item => item.company_name === company)
+            .map(item => item.donation)
+            .reduce((total, item) => total + item))
+);
```

Comment:
Here is an example of another approach to this problem:
```javascript
const donationsByCompany = runners
  .reduce((acc, { company, donation }) => {
    if (acc.hasOwnProperty(company)) {
      acc[company] += donation;
      return acc;
    } else {
      return { ...acc, [company]: donation };
    }
  }, {});
```
using [object destructuring](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment#Object_destructuring).